[docs] integration card for Polars, Patito, misc icons updates

danielgafni opened the pull request:
Adding integration cards for Polars and Patito. 

The Patito integration has been added in https://github.com/dagster-io/community-integrations/pull/179

## Changelog

[docs] New integration pages for [Polars](https://docs.dagster.io/integrations/libraries/polars) and [Patito](https://docs.dagster.io/integrations/libraries/patito).

danielgafni commented:
cc @OwenKephart I'm really happy by the `TypeRouter` thing - it was easy to add support to another type, and derived types like `Optional[MyModel.DataFrame]` should work out of the box!

danielgafni commented on python_modules/libraries/dagster-polars/dagster_polars/io_managers/type_routers.py:
```diff
@@ -143,7 +143,51 @@ def is_base_type(self) -> bool:
+
+    @property
+    def inner_type(self) -> Any:
+        if issubclass(self.typing_type, pl.DataFrame):
+            return pl.DataFrame
+        elif issubclass(self.typing_type, pl.LazyFrame):
+            return pl.LazyFrame
```
what the hell Graphite

ion-elgreco commented:
In my previous job I actually created a patitio_model_to_dagster_type, so dagster does the type checking as part of the type validation.

danielgafni commented:
I think both options can be available. Or maybe validation can be removed from the IOManager once I implement dagster types for Patito

gfvioli commented:
I got notified by @ion-elgreco that this was requested. I'll put this on my to-do list to get this into the OSS.

cmpadden commented:
@danielgafni can you rebase off of master, we landed a fix to the `setuptools` change that broke everyone.

danielgafni commented:
@cmpadden done

danielgafni commented:
@gfvioli oh great, thank you! Open to that

cmpadden commented:
@danielgafni looks like we have a couple of pyright errors:

```
[2025-03-24T19:52:10Z] Creating temporary pyright config file at pyrightconfig-alt-1.json
[2025-03-24T19:52:10Z] Running pyright for environment `alt-1`...
[2025-03-24T19:52:10Z]   pyright --project=pyrightconfig-alt-1.json --outputjson --level=warning --warnings
[2025-03-24T19:52:23Z] Creating temporary pyright config file at pyrightconfig-master.json
[2025-03-24T19:52:23Z] Running pyright for environment `master`...
[2025-03-24T19:52:23Z]   pyright --project=pyrightconfig-master.json --outputjson --level=warning --warnings
[2025-03-24T20:00:02Z]
[2025-03-24T20:00:02Z] /workdir/python_modules/libraries/dagster-polars/dagster_polars/io_managers/type_routers.py:
[2025-03-24T20:00:02Z]   151:15: Import "patito" could not be resolved (reportMissingImports)
[2025-03-24T20:00:02Z]
[2025-03-24T20:00:02Z]   165:15: Import "patito" could not be resolved (reportMissingImports)
[2025-03-24T20:00:02Z]
[2025-03-24T20:00:02Z] /workdir/python_modules/libraries/dagster-polars/dagster_polars_tests/test_patito.py:
[2025-03-24T20:00:02Z]   2:7: Import "patito" could not be resolved (reportMissingImports)
[2025-03-24T20:00:02Z]
[2025-03-24T20:00:02Z] pyright 1.1.379
```

danielgafni commented:
I made some decisions in order to make type checks work with IOManagers seamlessly. I refactored @gfvioli's code a little so that it can be used in this context.

I don't see any benefits of the Dagster Type given that the IOManager can do the same work, and specifying a type annotation is more elegant than constructing a Dagster Type, passing it around, and probably still specifying the type annotation anyway! 

I decided to make these components work independently. The current design is: 
- `IOManager`s in `dagster-polars` recognize Patito models in type annotations and trigger data validation 
- Alternatively, a Dagster Type can be constructed from a Patito model. This type will work with any IOManagers and type annotations.

@cmpadden I (finally) added an integration card for Polars, and also one for Patito. They reference `.svg` files which are not here yet, could you please apply some SVG wizardry so we can get them here? `patito.svg` and `polars.svg`. Haven't addressed CI issues yet, but maybe it's a good idea to rerun it anyway  

@deepyaman curious to hear your thoughts on this.

danielgafni commented:
@cmpadden I pushed pyright pins

danielgafni commented on docs/docs/integrations/libraries/polars.md:
```diff
@@ -0,0 +1,64 @@
+
+The `dagster-polars` integration allows using Polars eager or lazy DataFrames as inputs and outputs with Dagster’s assets and ops. Type annotations are used to control whether to load an eager or lazy DataFrame. Lazy DataFrames can be sinked as output. Multiple serialization formats (Parquet, Delta Lake, BigQuery) and filesystems (local, S3, GCS, …) are supported.
+
+`dagster-polars` integrates with [Patito](https://github.com/JakobGM/patito) for data validation. Learn mode on the [Patito integration page](/integrations/patito).
```
Thanks you are the best

danielgafni commented:
@cmpadden please kick it again :pray:

danielgafni commented:
What's up with API docs?

```
› yarn build-api-docs
command not found: ./scripts/build-api-docs.sh
```

cmpadden commented:
@danielgafni a small one:
```
[2025-03-25T15:54:20Z] Running pyright for environment `master`...
[2025-03-25T15:54:20Z]   pyright --project=pyrightconfig-master.json --outputjson --level=warning --warnings
[2025-03-25T16:01:58Z]
[2025-03-25T16:01:58Z] /workdir/python_modules/libraries/dagster-polars/dagster_polars/io_managers/type_routers.py:
[2025-03-25T16:01:58Z]   164:65: Unnecessary "# type: ignore" comment (reportUnnecessaryTypeIgnoreComment)
[2025-03-25T16:01:58Z]
[2025-03-25T16:01:58Z] pyright 1.1.379
[2025-03-25T16:01:58Z] Finished in 465.423 seconds
```

cmpadden commented:
> What's up with API docs?
> 
> ```
> › yarn build-api-docs
> command not found: ./scripts/build-api-docs.sh
> ```

@danielgafni this is working fine for me--could it be a shell specific issue? I'm using Bash FWIW.

danielgafni commented:
I am on `zsh`. I don't think it matters. 

`yarn build-api-docs` is failing for me, not sure why. I can run the script manually, but: 

```
› ./scripts/build-api-docs.sh
zsh: ./scripts/build-api-docs.sh: bad interpreter: /bin/bash: no such file or directory
```

Uh ok. Classic mutable OS users assuming I am going to have programs randomly scattered around my filesystem in some magic locations. 

 Changing `/bin/bash` to `env bash` to be more inclusive towards non-traditional distro users.

cmpadden commented on docs/scripts/build-api-docs.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!env bash
```
This doesn't actually work for my system.

```
$ yarn build-api-docs
command not found: ./scripts/build-api-docs.sh
```

But `/usr/bin/env` works...

```suggestion
#!/usr/bin/env bash
```

danielgafni replied:
:rofl: will do that

danielgafni commented:
The integration pages are looking ok (except the svg are missing). 
But the API docs don't appear for me (even after `build-api-docs.sh)`: 

![image](https://github.com/user-attachments/assets/d4d9d7ac-f31d-4f20-a7b3-39f834b5806d)

cmpadden commented:
Pushed a commit to add these logos. Thoughts on using the Polars Rust logo?
<img width="288" alt="image" src="https://github.com/user-attachments/assets/5a8cb046-9c02-4b0e-aab1-652c4f725cbb" />

Found it here, but can easily change it to their normal one.
https://github.com/pola-rs/polars-static/blob/master/web/polars-logo-rust.svg

danielgafni commented:
I think we should use their normal logo. In fact, we are writing :snake: :)

cmpadden commented:
> I think we should use their normal logo. In fact, we are writing 🐍 :)

ok! updated to their `polar-logos-python.svg`!

danielgafni commented:
1. Added more tests
2. I came up with [this](https://github.com/dagster-io/dagster/pull/28690/files#diff-da983a32fd4378ffc9e00f955cd3640837f28347a6bb19033b99fb01e0c665bcR186) hack which prevents us from running data validation twice if both the IOManager from `dagster-polars` and a Patito `DagsterType` are used at the same time. This is because data validation can be expensive. I think this exposes a slight limitation of the current `DagsterType` + `IOManager` system: both can do very similar things, but lack a good way to communicate with each other. I will create a Dicsussion about that. cc @smackesey @deepyaman 
3. I added a `ruff.toml` file to ban top-level `patito` imports, and looks like this triggered a `ruff` reformat of `dagster-polars`. Apparently previously it hasn't been fully discovered by Ruff! This is concerning, looks like other libraries are not discovered too. I will check that later. @cmpadden sorry for the huge diff, but I didn't have a good way to rollback this `ruff` invocation (lots of my actual changes were uncommitted).

danielgafni commented:
I believe this PR should be ready for review

cmpadden commented:
@danielgafni - seeing this error in the "check manifest" step:
```
[2025-03-26T14:13:21Z] + check-manifest python_modules/libraries/dagster-polars
[2025-03-26T14:13:22Z] lists of files in version control and sdist do not match!
[2025-03-26T14:13:22Z] missing from sdist:
[2025-03-26T14:13:22Z]   ruff.toml
[2025-03-26T14:13:22Z] suggested MANIFEST.in rules:
[2025-03-26T14:13:22Z]   include *.toml
```
https://github.com/dagster-io/dagster/blob/master/.buildkite/dagster-buildkite/dagster_buildkite/steps/dagster.py#L145

danielgafni commented:
@cmpadden fixed

cmpadden commented:
> @cmpadden fixed

Looks like we're running green. 👍

danielgafni commented on python_modules/libraries/dagster-polars/dagster_polars/io_managers/type_routers.py:
```diff
@@ -143,16 +166,85 @@ def is_base_type(self) -> bool:
+        return False
+
+    @property
+    def requires_data_validation(self) -> bool:
```
https://github.com/dagster-io/dagster/pull/28690#issuecomment-2754417141

deepyaman commented:
High-level thoughts:

- In general, as far as "Dagster-maintained" integrations go, I would rather support a data validation library (like Pandera) that can handle more than just Polars. While this may provide a good experience for Polars users, you still have the issue of what to do with Spark, pandas, and other library users (which is important, since data validation should be a core capability IMO).
- On the issues supporting Pandera:

    > Since it can dynamically generate model-specific DataFrame and LazyFrame classes, we can use it for type annotations in Dagster code right now (https://github.com/dagster-io/dagster/issues/22694 are not supported by Dagster at the moment). This makes Patito a good option for seamless data validation based on type annotations (in line with the rest of dagster-polars), unlike pandera, which uses generics.
    
    Could something akin to `patito_model_to_dagster_type` for Pandera circumvent these issues? (Actually, reading more, I think you mention in places like https://github.com/dagster-io/dagster/issues/22694 that yes, this is the necessary workaround.) Alternatively, maybe there is a way to even support an alternative syntax for annotating using Pandera that works with Dagster Types.

- Re type routers:
  - It would be good to support this in a unified way across Dagster (if the functionality should be supported at all? which is a [bigger question](https://github.com/dagster-io/community-integrations/pull/143)... )
  - Out of curiosity, did you consider using something built-in like https://docs.python.org/3/library/functools.html#functools.singledispatch? Do you think it would be sufficient? My same thought around the type handler implementation in DB I/O manager was, why wasn't it implemented using something like this.

- I think the hack to stop double validation is fair, but I also agree with your comment (I believe, outside of this thread), that it would be good to be able to emit asset check results and not be triggering this through type handlers?

I'm curious—is the main reason for going with Patito here that you can avoid the annotation issues with Pandera (in which case, I think we should prioritize solving that?) or because Patito is a better Polars validation library? In the end, I think ideally we should unblock the thing that is providing a better experience for users across Dagster, for those using just Polars and also for those using other libraries.

deepyaman commented:
> Patito is simply much better than pandera.

Why? I'm not familiar with Patito and haven't used Polars much, so this is for my edification.

> Also patito_model_to_dagster_type equivalent already exists for pandera, see dagster-pandera

Right, thanks!

danielgafni commented:
> Why? I'm not familiar with Patito and haven't used Polars much, so this is for my edification.

Patito is not just a validation lib, it also provides some handy method for working with the model class as values. For example: 

```py
for item in df.iter_models():
    item.my_field 
```

There are other methods which allow to extract model instances from the DataFrame, and has handy method like `df.cast(OtherModel)`.

Also, the Patito models are just Pydantic models, so they integrate well with other code that uses pydantic.  

To my understanding, Pandera has more advanced features (don't have experience with it so this is based on my intuition) around data validation and synthesis, but Patito has better APIs for more regular tasks. 

I have been using Patito at my current job and I do like it so far. Again, I don't have experience with Pandera, it might be that Pandera is great too. 

But I think we can just support both given that the `TypeRouter` pattern makes it very easy to add support for new types and this code is very unlikely to change in the future. 

I think we could just also take the `TypeRouter` list as an argument to the IOManager - in this case we can take patito-specific code out of the default IOManager, and allow the user to add an additional `PatitoTypeRouter` when instantiating the IOManager. I really like this composability option, but I also think that we can just have `PatitoTypeRouter` included by default.

danielgafni commented:
> In general, as far as "Dagster-maintained" integrations go, I would rather support a data validation library (like Pandera) that can handle more than just Polars. While this may provide a good experience for Polars users, you still have the issue of what to do with Spark, pandas, and other library users (which is important, since data validation should be a core capability IMO).

I think this will become possible once all the IOManagers are refactored to take `TypeRouter`/`TypeHandler` as constructor argument (given we decide to do it)

danielgafni commented:
Another solution on the `IOManager`/`DagsterType` duplication would be giving the `IOManager` control over `DagsterType` instantiation. 

Currently, we always have an implicit `DagsterType` created from the type annotation if the `dagster_type` argument was unset. 

I could see the `IOManager` having more control over the implicit `DagsterType` creation process (perhaps via the corresponding `TypeRouter`). This would be ideal as we could only do data validation in the `DagsterType`, and than it wouldn't matter if it was constructed from the type annotation or provided explicitly by the user.

gfvioli commented:
Thank you @danielgafni  for the help in making the DagsterType convertor ready to merge into `dagster-polars`. 

Always happy to help!

danielgafni commented:
Hi @deepyaman, how is it going?

deepyaman commented:
Sorry for the delay here. Unfortunately, it will unnecessarily block progress here to wait for an answer, and it makes more sense for `dagster-polars` to live in `community-integrations`. I will make a PR to move it over, and then you can re-raise this here (and bring it in).
